### PR TITLE
Pulling out a base class from benchmark gym game mode that can be used for offloading

### PR DESCRIPTION
**Pull request by MatthewSandfordImprobable**
All logic that concerns NFR metrics has been pulled out into a separate game mode base class. 
The offloading benchmark game mode now uses this new game mode as it's parent. This allows the offloading benchmark gym to take advantage of the UX metric. 

The offloading characater and simplayer BPs have had the UX and UXReporter component added to them. 

The game state now has two counter components counting the number of NPCs and SimPlayers. 

The offloading game mode now use the same worker flags and command line args keys as the code. "num_npcs" has now been update to "total_npcs" to reflect the benchmark gym.

### CODE REVIEW

**Comment by MatthewSandfordImprobable** [comment]
Most of the commits in the commit history are already in master. Not sure why they show up here...

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 1]
```diff
@@ -298,54 +155,34 @@ bool ABenchmarkGymGameMode::ShouldUseCustomSpawning()
...
-		}
-		// Set default value of PlayerDensity equal to TotalPlayers. Will be overwritten if PlayerDensity option is specified.
-		PlayerDensity = ExpectedPlayers;
-		if (NetDriver != nullptr && NetDriver->SpatialWorkerFlags != nullptr)
+
+		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Using worker flags to load custom spawning parameters."));
+		FString PlayerDensityString;
+
+		const USpatialWorkerFlags* SpatialWorkerFlags = NetDriver != nullptr ? NetDriver->SpatialWorkerFlags : nullptr;
+		if (SpatialWorkerFlags != nullptr)
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:175]: Could be worth adding a check here, we require these to be set or the test will be invalid
```suggestion
		checkf(SpatialWorkerFlags, "Worker flags are requried to be set");
		if (SpatialWorkerFlags != nullptr)
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.h [hunk 2]
```diff
@@ -4,20 +4,19 @@
...
+DECLARE_LOG_CATEGORY_EXTERN(LogBenchmarkGymGameMode, Log, All);
 
 typedef TPair<TWeakObjectPtr<AController>, int> ControllerIntegerPair;
 
 /**
  *
  */
 UCLASS()
-class GDKTESTGYMS_API ABenchmarkGymGameMode : public AGameModeBase
+class GDKTESTGYMS_API ABenchmarkGymGameMode : public ABenchmarkGymGameModeBase
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.h:19]: Maybe ABenchmarkGymGameModeBase could be ANFRGymGameModeBase or something?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+#include "Engine/World.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "GDKTestGymsGameInstance.h"
+#include "Interop/SpatialWorkerFlags.h"
+#include "Misc/CommandLine.h"
+#include "Utils/SpatialMetrics.h"
+
+DEFINE_LOG_CATEGORY(LogBenchmarkGymGameModeBase);
+
+const FString ABenchmarkGymGameModeBase::TotalPlayerWorkerFlag = TEXT("total_players");
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:14]: Could put these in the anon namespace if the blueprint functions were implemented in the cpp, although can see why it's done like this too so just a thought

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+	// These values need to match the GDK scenario validation equivalents
+	, MinAcceptableFPS(20.0f) // Same for both client and server currently
+	, MinDelayFPS(120.0f)
+	, ActivePlayers(0)
+{
+	PrimaryActorTick.bCanEverTick = true;
+	SecondsTillPlayerCheck = 15.0f * 60.0f;
+
+	// Seamless Travel is not currently supported in SpatialOS [UNR-897]
+	bUseSeamlessTravel = false;
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:50]: Do we use this?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	, bPlayersHaveJoined(false)
+	, bHasUxFailed(false)
+	, bHasFpsFailed(false)
+	// These values need to match the GDK scenario validation equivalents
+	, MinAcceptableFPS(20.0f) // Same for both client and server currently
+	, MinDelayFPS(120.0f)
+	, ActivePlayers(0)
+{
+	PrimaryActorTick.bCanEverTick = true;
+	SecondsTillPlayerCheck = 15.0f * 60.0f;
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:47]: Can also be in the initializer list - same with seamlessTravel below

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 4 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:50]: Doesn't seem so, I just kept it from what was there.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+	// Seamless Travel is not currently supported in SpatialOS [UNR-897]
+	bUseSeamlessTravel = false;
+}
+
+void ABenchmarkGymGameModeBase::BeginPlay()
+{
+	Super::BeginPlay();
+
+	ParsePassedValues();
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:57]: This was delayed until FindPlayerStart for some strange reasons, I can't remember exactly why now (possibly WorkerFlags not being available) but worth running the NFR tests over this!

Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:57]: Yup running them now!

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 5 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:47]: SecondsTillPlayerCheck can be actually. bUseSeamlessTravel apparently not.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 3 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:14]: Yeah either or. I don't mind personally.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 4 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:50]: Ah might as well remove, it was probably some legacy code that has gone now!

**Comment by m-samiec** [comment]
> The game state now has two counter components counting the number of NPCs and SimPlayers.

Two components? Or one component counting both?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+DECLARE_LOG_CATEGORY_EXTERN(LogBenchmarkGymGameModeBase, Log, All);
+
+UCLASS()
+class GDKTESTGYMS_API ABenchmarkGymGameModeBase : public AGameModeBase
+{
+	GENERATED_BODY()
+public:
+	ABenchmarkGymGameModeBase();
+
+	UFUNCTION(BlueprintCallable)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h:20]: Why are these exposed?

File: spatial/offloading_test.json [hunk 8]
```diff
@@ -30,17 +30,7 @@
         "layer": "UnrealWorker",
         "rectangle_grid": {
```
Comment (m-samiec) [inline spatial/offloading_test.json:31]: Do we need the rectangle_grid? Can't we just specify 2 workers?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+		PrintUXMetric = 10.0f;
+#if !WITH_EDITOR
+		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("UX metric values. RTT: %.8f, ViewLateness: %.8f, ActivePlayers: %d"), AveragedClientRTTSeconds, AveragedClientViewLatenessSeconds, ActivePlayers);
+#endif
+	}
+}
+
+void ABenchmarkGymGameModeBase::ParsePassedValues()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("TotalPlayers")))
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:203]: Can we reuse the same "OverrideSpawning" arg used elsewhere?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+			}
+
+			if (NetDriver->SpatialWorkerFlags->GetWorkerFlag(MaxLatenessWorkerFlag, MaxViewLateness))
+			{
+				MaxClientViewLatenessSeconds = FCString::Atoi(*MaxViewLateness);
+			}
+		}
+	}
+
+	UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Players %d, NPCs %d"), ExpectedPlayers, TotalNPCs);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:246]: Realise you copied these - but can you output the value of everything possibly set in this file at the bottom here. Useful for debugging.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 11]
```diff
@@ -16,17 +16,13 @@
...
-{
-	const FString AverageClientRTTMetricName = TEXT("UnrealAverageClientRTT");
-	const FString AverageClientViewLatenessMetricName = TEXT("UnrealAverageClientViewLateness");
-	const FString PlayersSpawnedMetricName = TEXT("UnrealActivePlayers");
-}
+DEFINE_LOG_CATEGORY(LogBenchmarkGymGameMode);
 
 ABenchmarkGymGameMode::ABenchmarkGymGameMode()
+	: bInitializedCustomSpawnParameters(false)
+	, NumPlayerClusters(4)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:23]: driveby - but this should really default to 1.

**Comment by m-samiec** [comment]
Looks good - could we go over the blueprint changes also?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h (hunk 7 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.h:20]: We had functionality in the Offloading gym game mode to dynamically update when the worker flag value changes. I wanted to ensure we use the same keys. I might take a look at the logic inside offloading game mode and see if it can be simplified.

File: spatial/offloading_test.json (hunk 8 above)
Comment (MatthewSandfordImprobable) [reply spatial/offloading_test.json:31]: Ahh, if that can be done... perhaps! I just went on recommendation from Jen.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 9 above)
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:203]: We could but I don't think OverrideSpawning is relevant for offloading.

**Comment by MatthewSandfordImprobable** [comment]
> > The game state now has two counter components counting the number of NPCs and SimPlayers.
> 
> Two components? Or one component counting both?

It did have two components but I have changed it to one. Didn't realize it was an array of actors.

File: spatial/offloading_test.json (hunk 8 above)
Comment (m-samiec) [reply spatial/offloading_test.json:31]: `entity_id_sharding` or `hex_grid` allow you to just specify your number of workers in one dimension. @jennifer-at-improbable-io what do you think?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 9 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp]: I guess I don't like the inference that `TotalPlayers` means you should read the rest of the values from command line. Can we implement a process that can be shared between setups then?

Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp]: Sure so we change OverrideSpawning to something more generic like "ParseCommandLine"?

Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp]: In fact we should have both.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 12]
```diff
@@ -298,54 +151,34 @@ bool ABenchmarkGymGameMode::ShouldUseCustomSpawning()
...
-		FParse::Value(FCommandLine::Get(), TEXT("MaxRoundTrip="), MaxClientRoundTripSeconds);
-		FParse::Value(FCommandLine::Get(), TEXT("MaxLateness="), MaxClientViewLatenessSeconds);
+		FParse::Value(*CommandLine, TEXT("PlayerDensity="), PlayerDensity);
 	}
-	else
+	else if(GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
 	{
-		UE_LOG(LogBenchmarkGym, Log, TEXT("Using worker flags to load custom spawning parameters."));
-		FString TotalPlayersString, PlayerDensityString, TotalNPCsString, MaxRoundTrip, MaxViewLateness;
+		USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GetNetDriver());
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:165]: If you boot in editor in standalone, client-only mode, does this crash?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 13]
```diff
@@ -298,54 +151,34 @@ bool ABenchmarkGymGameMode::ShouldUseCustomSpawning()
...
-			ExpectedPlayers = FCString::Atoi(*TotalPlayersString);
-		}
-		// Set default value of PlayerDensity equal to TotalPlayers. Will be overwritten if PlayerDensity option is specified.
-		PlayerDensity = ExpectedPlayers;
-		if (NetDriver != nullptr && NetDriver->SpatialWorkerFlags != nullptr)
+
+		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Using worker flags to load custom spawning parameters."));
+		FString PlayerDensityString;
+
+		const USpatialWorkerFlags* SpatialWorkerFlags = NetDriver != nullptr ? NetDriver->SpatialWorkerFlags : nullptr;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:171]: Where'd in a point where NetDriver has already been checked, can simplify this.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+			FOnWorkerFlagsUpdatedBP WorkerFlagDelegate;
+			WorkerFlagDelegate.BindDynamic(this, &ABenchmarkGymGameModeBase::OnWorkerFlagUpdated);
+			SpatialWorkerFlags->BindToOnWorkerFlagsUpdated(WorkerFlagDelegate);
+		}
+	}
+
+	UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Players %d, NPCs %d, RoundTrip %d, ViewLateness %d"), ExpectedPlayers, TotalNPCs, MaxClientRoundTripSeconds, MaxClientViewLatenessSeconds);
+}
+
+void ABenchmarkGymGameModeBase::OnWorkerFlagUpdated(const FString& FlagName, const FString& FlagValue)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:312]: Didn't we decide that we didn't need this functionality?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+	{
+		MaxClientViewLatenessSeconds = FCString::Atoi(*FlagValue);
+	}
+
+	UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Worker flag updated - Flag %s, Value %s"), *FlagName, *FlagValue);
+}
+
+void ABenchmarkGymGameModeBase::SetTotalNPCs(int32 Value)
+{
+	TotalNPCs = Value;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:336]: Do you only want to call `OnTotalNPCsUpdated` if this value is different? There may be a disconnect otherwise as the non-auth server won't have `OnRepTotalNPCs` called on it if the values haven't changed.

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+} // anonymous namespace
+
+FString ABenchmarkGymGameModeBase::ReadFromCommandLineKey = TEXT("ReadFromCommandLine");
+
+ABenchmarkGymGameModeBase::ABenchmarkGymGameModeBase()
+	: ExpectedPlayers(1)
+	, SecondsTillPlayerCheck(15.0f * 60.0f)
+	, PrintUXMetric(10.0f)
+	, MaxClientRoundTripSeconds(150)
+	, MaxClientViewLatenessSeconds(150)
```
[the remark this answers is not in the document]
Comment (MatthewSandfordImprobable) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:45]: Always been 150. Previously we had in the game mode: 

MaxClientRoundTripSeconds = MaxClientViewLatenessSeconds = 150;	

 I can change it to 300 though?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 17]
```diff
@@ -323,54 +151,35 @@ bool ABenchmarkGymGameMode::ShouldUseCustomSpawning()
...
-			ExpectedPlayers = FCString::Atoi(*TotalPlayersString);
-		}
-		// Set default value of PlayerDensity equal to TotalPlayers. Will be overwritten if PlayerDensity option is specified.
-		PlayerDensity = ExpectedPlayers;
-		if (NetDriver != nullptr && NetDriver->SpatialWorkerFlags != nullptr)
+		UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Using worker flags to load custom spawning parameters."));
+		FString PlayerDensityString;
+
+		USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GetNetDriver());
+		if (ensure(NetDriver != nullptr))
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:169]: Nice!

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:45]: It's set to 300 in all out tests - lets change it here also

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+			{
+				bHasFpsFailed = true;
+#if !WITH_EDITOR
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("FPS check failed. FPS: %.8f"), FPS);
+#endif
+			}
+		}
+	}
+
+	bool bClientFpsWasValid = true;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:207]: From here on this only needs to be done on the auth server. Perhaps split this into two functions, one for server (auth and non-auth), and one for client checks (auth only)?
